### PR TITLE
Expand `reduceBool` and add `reduceSlice`

### LINKED ISSUE #195: Support more boolean operations in constant evaluation
[`reduceBool`](https://github.com/tact-lang/tact/blob/cd34466ceb552977518ce60af10efad6f06ce69b/src/types/resolveConstantValue.ts#L62) is responsible for evaluating Tact constants in compile-time. Currently it supports only `&&` and `||` for booleans, and there is the TODO comment to add more operations: https://github.com/tact-lang/tact/blob/cd34466ceb552977518ce60af10efad6f06ce69b/src/types/resolveConstantValue.ts#L77

The following boolean operations could be easily added: `==` and `!=`.

We could also consider adding the following operations between integer constants that results in boolean:  `!=`, `>`, `<`, `>=`, `<=`, `==`.

In general, this feature is quite useful, as it allows the creation of constants with meaningful names instead of "magic values" in code, which is considered a best practice in programming. However, the need for these specific cases is quite rare.

Found while working on #133.

### LINKED ISSUE #195: Change all `map<k, v>` occurences to `map<K, V>` to conform to `grammar.ohm`

### PR DESCRIPTION
**Pull request by Gusarich**
resolves tact-lang/tact-docs#195 

I didn't cover that with new tests yet, going to do it soon

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -14,6 +14,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Changed
 - Update the `dump` function to handle addresses: PR [#175](https://github.com/tact-lang/tact/pull/175)
 - The implicit empty `init` function is now present by default in the contract if not declared: PR [#167](https://github.com/tact-lang/tact/pull/167)
+- Improved `Bool` reduction in constant expressions: PR [#195](https://github.com/tact-lang/tact/pull/195)
```
Comment (byakuren-hijiri) [inline CHANGELOG.md]:
```suggestion
- Improved `Bool` reduction in constant expressions: PR [#197](https://github.com/tact-lang/tact/pull/197)
```

File: src/types/resolveConstantValue.ts [hunk 2]
```diff
@@ -59,22 +60,51 @@ function reduceInt(ast: ASTExpression): bigint {
...
+                    } else if (leftType.name === 'String' && rightType.name === 'String') {
+                        return reduceString(ast.left) === reduceString(ast.right);
+                    } else if (leftType.name === 'Int' && rightType.name === 'Int') {
+                        return reduceInt(ast.left) === reduceInt(ast.right);
+                    } else if (leftType.name === 'Bool' && rightType.name === 'Bool') {
+                        return reduceBool(ast.left, ctx) === reduceBool(ast.right, ctx);
+                    } else if (leftType.name === 'Slice' && rightType.name === 'Slice') {
+                        return reduceSlice(ast.left).asCell().equals(reduceSlice(ast.right).asCell());
+                    }
+                } else if (leftType.kind === 'null' && rightType.kind === 'null') {
```
Comment (byakuren-hijiri) [inline src/types/resolveConstantValue.ts:115]: What is the use case for this? Could you provide a minimal example please?

File: src/types/resolveConstantValue.ts [hunk 3]
```diff
@@ -59,22 +60,51 @@ function reduceInt(ast: ASTExpression): bigint {
...
+            const rightType = getExpType(ctx, ast.right);
+            if (ast.op === '>') {
+                return reduceInt(ast.left) > reduceInt(ast.right);
+            } else if (ast.op === '<') {
+                return reduceInt(ast.left) < reduceInt(ast.right);
+            } else if (ast.op === '>=') {
+                return reduceInt(ast.left) >= reduceInt(ast.right);
+            } else if (ast.op === '<=') {
+                return reduceInt(ast.left) <= reduceInt(ast.right);
+            } else if (ast.op === '==') {
```
Comment (byakuren-hijiri) [inline src/types/resolveConstantValue.ts:100]: How about `!=`?

File: src/types/resolveConstantValue.ts (hunk 2 above)
Comment (byakuren-hijiri) [reply src/types/resolveConstantValue.ts:115]: Ideally, we should extend the [`constants.tact`](https://github.com/tact-lang/tact/blob/7ef2ae17baef890d6489688d523bcdf77ce7b561/src/test/features/constants.tact) test

File: src/types/resolveConstantValue.ts [hunk 4]
```diff
@@ -126,6 +156,24 @@ function reduceCell(ast: ASTExpression): Cell {
     throwError('Cannot reduce expression to a constant Cell', ast.ref);
 }
 
+function reduceSlice(ast: ASTExpression): Slice {
+    if (ast.kind === 'op_static_call') {
```
Comment (byakuren-hijiri) [inline src/types/resolveConstantValue.ts]: I would suggest defining the `isSlice` function for clarity:

```typescript
function isSlice(ast: ASTExpression): bool {
    return (ast.kind === 'op_static_call') && (ast.name === 'slice') && (ast.args.length === 1);
}
```

File: src/types/resolveConstantValue.ts (hunk 2 above)
Comment (byakuren-hijiri) [reply src/types/resolveConstantValue.ts:115]: Ah, I see mention about the tests in the description.

File: CHANGELOG.md (hunk 1 above)
Comment (anton-trunov) [reply CHANGELOG.md]: Let's document here what exactly is allowed in constant expressions after this PR is merged (not only the new stuff, but everything) -- this is going to go into the docs, see https://github.com/tact-lang/tact-docs/issues/152

**Comment by Gusarich** [comment]
Current implementation doesn't actually work and throws errors on compiling new test cases.
I'll investigate the problem and try to fix it.

**Comment by anton-trunov** [comment]
The issue here is that constant evaluation is performed on a non-typechecked AST. So when you call `getExpType` to dispatch on the type of `==` / `!=` operators, there is no actual type information. The proper fix here would be to refactor the compiler to perform constant evaluation after typechecking (and to actually do this refactoring we need a new type describing typed ASTs).

**Comment by anton-trunov** [comment]
Looks like this PR should be closed now? Because the architecture of the constant evaluator has changed a lot in #432. Of course, we can still improve a lot of things in the constant evaluator, adding support for slices would be especially nice.

**Comment by Gusarich** [comment]
yes, right

**Pull request closed without merging**